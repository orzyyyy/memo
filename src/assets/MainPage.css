.MainPage img {
  width: 100%;
}

.MainPage .card {
  width: 25%;
  text-align: center;
  cursor: pointer;
}

.MainPage .animate-appear {
  animation: 0.3s appear 0s;
}

@keyframes appear {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
