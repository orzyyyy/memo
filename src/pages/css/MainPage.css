.MainPage img {
  width: 100%;
}

.MainPage {
  height: 100vh;
  min-height: 100vh;
  transition: background 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
}

.MainPage .list-item {
  cursor: pointer;
  transition: background 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
  height: 40px;
  line-height: 40px;
  padding-left: 8px;
}

.MainPage .list-item:hover {
  background: #e6f7ff;
  transition: background 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
}

.MainPage ul {
  margin-top: 8px;
  margin-bottom: 8px;
  list-style: none;
}

.MainPage li {
  margin-top: 8px;
  margin-bottom: 8px;
  transition: background 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
}

.MainPage .icon-apartment {
  -webkit-mask-image: url(../svg/apartment.svg);
  mask-image: url(../svg/apartment.svg);
  width: 20px;
  height: 20px;
  display: inline-block;
  margin-right: 10px;
}

.MainPage .icon-file-markdown {
  -webkit-mask-image: url(../svg/file-markdown.svg);
  mask-image: url(../svg/file-markdown.svg);
  width: 20px;
  height: 20px;
  display: inline-block;
  margin-right: 10px;
}
