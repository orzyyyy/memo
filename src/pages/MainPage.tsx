import React, { Component } from 'react';
import './css/MainPage.css';
import { Menu, Layout, Icon, Button, Input } from 'antd';
const { SubMenu } = Menu;
const { Content, Footer, Sider, Header } = Layout;
import { SiderProps } from '../controller/MainPageDataController';
import { MappingProps } from '../../server/controller/DocumentController';

export interface MainPageProps {
  dataSource: MappingProps[];
  onEdit: (dataItem?: any) => void;
  onDelete?: (dataItem: any) => void;
  menuData: SiderProps[];
  onExhentaiDownload: (event: React.KeyboardEvent<HTMLInputElement>) => void;
  renderContent: (
    props: MainPageProps,
    state: MainPageState,
    event?: any,
  ) => React.ReactNode;
  onMenuClick: (keyPath: string[]) => void;
}
export interface MainPageState {
  siderOpenKey: string;
  siderSelectedKey: string;
}

export default class MainPage extends Component<MainPageProps, MainPageState> {
  static getDerivedStateFromProps(
    prevProps: MainPageProps,
    prevState: MainPageState,
  ) {
    const { menuData } = prevProps;
    const { siderSelectedKey } = prevState;
    if (menuData.length > 0 && !siderSelectedKey) {
      return {
        siderOpenKey: menuData[0].title,
        siderSelectedKey: menuData[0].title,
      };
    }
    return null;
  }

  state: MainPageState = {
    siderOpenKey: '',
    siderSelectedKey: '',
  };

  componentDidMount() {
    window.onresize = () => this.setState({});
  }

  handleMenuClick = ({ keyPath }: { keyPath: string[] }) => {
    this.props.onMenuClick(keyPath);
    this.setState({
      siderOpenKey: keyPath[1],
      siderSelectedKey: keyPath[0],
    });
  };

  renderSider = () => {
    const { menuData } = this.props;
    const { siderSelectedKey } = this.state;
    return (
      <Sider theme="light">
        <Menu
          selectedKeys={[siderSelectedKey]}
          mode="inline"
          onClick={this.handleMenuClick}
        >
          {menuData.map((item: any) => {
            const { key, title, children } = item;
            if (!children) {
              return <Menu.Item key={key}>{title}</Menu.Item>;
            }
            return (
              <SubMenu key={key} title={title}>
                {children.map((jtem: any) => (
                  <Menu.Item key={jtem.key}>{jtem.value}</Menu.Item>
                ))}
              </SubMenu>
            );
          })}
        </Menu>
      </Sider>
    );
  };

  handleListItemClick = ({
    category,
    id,
  }: {
    category: 'mapping' | 'markdown';
    id: string;
  }) => {
    location.hash = `/${category}/${id}`;
  };

  renderContent = () => {
    const wrapperHeight = document.body.clientHeight - 48 - 90;
    return (
      <Content style={{ marginLeft: 8 }}>
        <div
          style={{
            padding: '8px 16px 8px 8px',
            background: '#fff',
            height: wrapperHeight,
            overflow: 'auto',
          }}
        >
          {this.props.renderContent(this.props, this.state, this)}
        </div>
      </Content>
    );
  };

  renderFooter = () => (
    <Footer style={{ textAlign: 'center' }}>
      <div>
        你睡了一下午，醒的时候屋子里黑漆漆、一点声音都没有。抬头望望窗外，天还没完全黑。四处摸了摸，在枕头下找到手机。打开后屏幕亮起，干净，没有一条消息
      </div>
      <div>
        打开电脑，打开 github。pull request 写得很菜，连 core
        都在喷你，但忽然就不孤独了
      </div>
    </Footer>
  );

  renderHeader = () => {
    const { onExhentaiDownload, onEdit } = this.props;
    return (
      <Header style={{ background: 'rgba(0, 0, 0, 0)', height: 48 }}>
        <Input
          onPressEnter={onExhentaiDownload}
          style={{ position: 'absolute', right: 80, top: 10, width: 350 }}
        />
        <Button
          style={{ position: 'absolute', right: 24, top: 10 }}
          onClick={() => onEdit()}
        >
          <Icon type="plus" />
        </Button>
      </Header>
    );
  };

  render = () => (
    <Layout className="MainPage">
      {this.renderSider()}
      <Layout>
        {this.renderHeader()}
        {this.renderContent()}
        {this.renderFooter()}
      </Layout>
    </Layout>
  );
}
