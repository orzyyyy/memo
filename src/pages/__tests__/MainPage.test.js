import React from 'react';
import { shallow, mount } from 'enzyme';
import MainPage from '../MainPage';

const menuData = [
  {
    key: 'all',
    title: 'all',
  },
  {
    key: '读书',
    title: '读书',
    children: [
      {
        key: '摘抄',
        value: '摘抄',
      },
      {
        key: '思路',
        value: '思路',
      },
    ],
  },
];
const dataSource = [
  {
    id: '32e89f38116080d6b7608cf2137f3c96',
    title: '音程表',
    url: './assets/mapping/32e89f38116080d6b7608cf2137f3c96.json',
    createTime: 1559522075108,
    modifyTime: 1559523700429,
    type: '音乐',
    subType: '乐理',
    category: 'markdown',
  },
  {
    id: '7a73d21c13f1eb55053d4af66e0d057b',
    title: 'windows 下安装 rust',
    url: './assets/mapping/7a73d21c13f1eb55053d4af66e0d057b.json',
    createTime: 1559617097664,
    modifyTime: 1559633639037,
    type: 'Rust',
    subType: '安装',
    category: 'markdown',
  },
];

describe('MainPage', () => {
  it('render correctly', () => {
    const wrapper = shallow(
      <MainPage dataSource={dataSource} menuData={menuData} onEdit={null} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('onMenuClick works', async () => {
    const onMenuClick = jest.fn();
    const wrapper = await mount(
      <MainPage
        dataSource={dataSource}
        menuData={menuData}
        onMenuClick={onMenuClick}
      />,
    );
    wrapper
      .find('MenuItem')
      .at(1)
      .simulate('click');
    expect(onMenuClick).toHaveBeenCalled();
  });

  it('import Header dynamically correctly', async () => {
    const wrapper = await mount(
      <MainPage dataSource={dataSource} menuData={[]} />,
    );
    wrapper.setProps({ isLocal: true });
    await wrapper.update();
    expect(wrapper.state('DynamicHeader')).toBeNull();
  });
});
