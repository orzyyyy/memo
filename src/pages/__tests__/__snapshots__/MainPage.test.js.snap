// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage render correctly 1`] = `
<Adapter
  className="MainPage"
>
  <Sider
    theme="light"
  >
    <Menu
      mode="inline"
      onClick={[Function]}
      selectedKeys={
        Array [
          "all",
        ]
      }
    >
      <MenuItem
        key="all"
      >
        all
      </MenuItem>
      <SubMenu
        key="读书"
        title="读书"
      >
        <MenuItem
          key="摘抄"
        >
          摘抄
        </MenuItem>
        <MenuItem
          key="思路"
        >
          思路
        </MenuItem>
      </SubMenu>
    </Menu>
  </Sider>
  <Adapter>
    <Adapter
      style={
        Object {
          "background": "rgba(0, 0, 0, 0)",
          "height": 48,
        }
      }
    >
      <Input
        disabled={true}
        style={
          Object {
            "position": "absolute",
            "right": 80,
            "top": 10,
            "width": 350,
          }
        }
        type="text"
      />
      <Button
        block={false}
        disabled={true}
        ghost={false}
        htmlType="button"
        loading={false}
        onClick={[Function]}
        style={
          Object {
            "position": "absolute",
            "right": 24,
            "top": 10,
          }
        }
      >
        <Icon
          type="plus"
        />
      </Button>
    </Adapter>
    <Adapter
      style={
        Object {
          "marginLeft": 8,
        }
      }
    >
      <div
        className="main-page-content-wrapper"
        style={
          Object {
            "background": "#fff",
            "height": -138,
            "overflow": "auto",
            "padding": "8px 16px 8px 8px",
          }
        }
      />
    </Adapter>
    <Adapter
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <div>
        你睡了一下午，醒的时候屋子里黑漆漆、一点声音都没有。抬头望望窗外，天还没完全黑。四处摸了摸，在枕头下找到手机。打开后屏幕亮起，干净，没有一条消息
      </div>
      <div>
        打开电脑，打开 github。pull request 写得很菜，连 core 都在喷你，但忽然就不孤独了
      </div>
    </Adapter>
  </Adapter>
</Adapter>
`;
